module/zfs/zvol.c

```diff
@@ -1012,6 +1021,97 @@ zvol_last_close(zvol_state_t *zv)
 	    !(zv->zv_flags & ZVOL_RDONLY))
 		txg_wait_synced(dmu_objset_pool(zv->zv_objset), 0);
 	(void) dmu_objset_evict_dbufs(zv->zv_objset);
+}
+
+/*
+ * Suspend the zvol for recv and rollback.
+ */
+int
+zvol_suspend(const char *name)
+{
+	zvol_state_t *zv;
+	int error = 0;
+
+retry:
+	mutex_enter(&zvol_state_lock);
+	zv = zvol_find_by_name(name);
+	if (zv == NULL) {
+		error = ENOENT;
+		goto out;
+	}
+
+	/* we cannot block in rw_enter if already suspended */
+	if (rw_owner(&zv->zv_suspend_lock) != NULL) {
```

Comment:
I didn't want to use rw_tryenter here because it could easily starve. As for the deadlock, it is exactly because zvol_request don't take zvol_state_lock, so it won't cause lock inversion. Every other that has zvol_state_lock need to use rw_tryenter.

Though, you are probably right reference counting could make things easier...